### ISSUE TITLE
No support for Adaptive parallel, chained payments

### ISSUE THREAD

**Issue by sudhirkelappan** [paypal/PayPal-iOS-SDK#9]
The older MPL for iOS seemed to have support for chained and parallel payments using adaptive payments.

This library does not seem to have support for it. 

Pls. clarify. 

Any plans to provide it ?

**Comment by balkrishanyadav** [comment]
hello in Paypal any way to use adaptive payment in any way.. in parallel mode payment to all ...

**Comment by arundevma** [comment]
unfortunately you have no other options to use other than MPL. Still waiting for paypal to integrate adaptive payments in IOS SDK. You can also try Stripe payment gateway. Stripe integrates flawlessly in IOS.

**Comment by derekdowling** [comment]
I'd also be interesting in using this. As of right now, neither Braintree or Stripe are available in Canada.

**Comment by jordanhammond** [comment]
Hi @dgoldman-ebay, it's almost a year since this thread has been active almost 1 year now.  Whilst I am not looking to use chained payments in ios, I am looking for some better tooling for adaptive payments in mobile web.  

This is an area that has been largely ignored by the old classic APIs from PayPal. The mobile web user experience for AP is horrible - almost non supported.  The hope is the rest APIs will make the world a better place again and solve world peace.  

Do you have an update on the progress of the rest api supporting parallel and chained payments (deferred too)?

I've recently converted recurring payments to rest subscriptions which was a good change.  Would love to do the same for chained payments and hopefully properly starting to support mobile web customers.

**Comment by dgoldman-ebay** [comment]
@jordanhammond you and I are on the same page in wanting the REST APIs to complete their support for advanced payments, including chained and parallel payments. Also that world peace thing.

I wish I had an update for you, but there are lots of sometimes-conflicting priorities inside PayPal, and therefore nobody can yet give me an estimated date for the REST APIs to finish their conquest of adaptive payments. :crying_cat_face: 

I think that this is one of those features where until it actually ships, nobody really knows whether it will take a few more weeks or many more months.

**Comment by derekdowling** [comment]
Yeah, as it stands the whole thing is a bloody mess. 18 different APIs all doing slightly different things and none of them seem to play nicely with each other. Or if they do there is virtually no documentation explaining how you might accomplish that.

**Comment by Gaudon** [comment]
My business model is heavily reliant on the support of this chained payment methodology, I hope we get some solid updates in the not so distant future!

**Comment by arundevma** [comment]
If the mini browser option or lightbox option could be supported in UIWebview, adaptive payments can be easily brought to mobile platforms (Atleast as a temporary solution until SDK supports it). A workflow similar to MEC. But both options are not supported in UIWebview in IOS. Closest thing I could get was mini browser option. But webview was not redirecting to success or failure URLs after transaction. For market place apps paying two fees for single item purchase (one fee by seller for amount he receives, other for commission amount) is not feasible. Most often for small commission amounts like 0.5$ major portion will be eaten up by PayPal as PayPal fee. In one of our apps that we developed for our client we were forced to use Stripe, as our app was mobile only.

**Comment by Gaudon** [comment]
So it's been over a year now with people awaiting this functionality, are we still awaiting updates on the rest api side of things?

I feel like I have no choice but to simply not use PayPal, how frustrating!

**Comment by garfieldmypet** [comment]
@dgoldman-ebay Do you have any update regarding IOS sdk supporting adaptive chained payment? Is this still a priority for PayPal?

**Comment by mamaral** [comment]
@garfieldmypet I wouldn't hold my breath.

**Comment by Gaudon** [comment]
I watch this GitHub every single day awaiting this functionality and it seems it just isn't coming.

**Comment by tomwhipple** [comment]
Unfortunately I can't add any good news here, but I once again will again reiterate all this feedback to the team responsible for that API.

**Comment by garfieldmypet** [comment]
@tomwhipple Thank you for the reply. We really hope that REST API and IOS SDK supports chained payments soon. Currently the classic Adaptive payment flow doesn't have a mobile web checkout UI interface and doesn't have very good conversion rate on mobile devices. With the new REST API and IOS SDK , I am sure the checkout flow will be more smooth. It has been 2 years in waiting, and we really hope that PayPal will put this as one of the top priorities.

**Comment by tapwork** [comment]
So no updates on Delayed Chain Payments. Currently we need to use the old PayPal SDK which won't work anymore with iOS 9 (since NSURLConnection has been deprecated)
Is there any way to use UIWebView? The old SDK is a big no option 'cause of missing NSURLSession support and it annoys the users with some timeout alerts

**Comment by braebot** [comment]
@tapwork, are you referring to [MPL](https://github.com/paypal/sdk-packages/tree/gh-pages/MPL)?  It should be relatively straightfoward to add support for iOS 9.  I created https://github.com/paypal/sdk-packages/issues/50 in the sdk-packages repo to track that work.

**Comment by garfieldmypet** [comment]
Hi, braebot,

I really really hope that PayPal pushes forward to support chained payment in both the REST SDK and IOS SDK. It is really very important to us. We check this thread almost everyday hoping that PayPal will support it. 

When we post here, the IOS SDK team says we need to wait for REST API to support chained payment first. And when we posted in REST API support, the ticket got closed and merged back to this thread.

Please PayPal, please support chained payment for REST API. It really has been a long time and as a developer , please understand our frustration.

I believe a lot of other developer that want this feature feels the same way too. At least let us know by when it will be available.

**Comment by caiob3** [comment]
Please PayPal, please support chained payment for REST API. It really has been a long time and as a developer , please understand our frustration.

**Comment by milsirhc** [comment]
Yea, really hope that chained payment for REST API will be available soon. Come on PayPal, we are all waiting for this for a looooong time and if this delayed any further, Stripe might be a good alternative with their Stripe Connect API.

**Comment by Blahartinger** [comment]
BlackBerry just partnered with PayPal to support payments via BBM. I imagine they aren't using the old MPL library on BB10 and that in fact they are using the new REST API for such types of p2p payments. If this is the case then I imagine we'll see a public release in the near future.

**Comment by amirpervaiz086** [comment]
Hi guys,
  I need to implement user to user payment using paypal in one of my app. I want sender to send direct money to other user account just like how taxi apps work. Customer can directly pay to driver without taking any profit of merchant. It can be done by adaptive payment. By reading above comments, it seems like adaptive payment is not supported by iOS SDK so can anyone of you shed light on this and tell me if it can be done in the backend using adaptive payments through webpage?

I would really appreciate your deep insights.

**Comment by Blahartinger** [comment]
@amirpervaiz086 - We've done something interesting utilizing MPL to achieve somethign similar to what you're talking about. It's an easy to setup and free sdk to wrap up this functionality. We call it our P2P sdk.

Applicable Links:
www.vouc.hr - see P2P SDK
https://moneyhole.vouc.hr/developer
https://github.com/VouchrApp/MoneyHoleClient

**Comment by garfieldmypet** [comment]
Hi, PayPal support,

Is there any update on PayPal IOS SDK natively supporting chained payment. With IOS9 coming out in September 2015, we are really looking forward for PayPal IOS SDK and Rest SDK to support chained payment. We want to integrate with PayPal new API, but we need chained payment.

Do you have any timeline for this feature and is anyone in PayPal actually working on it? As for now, we can't use Braintree because even Braintree doesn't support Chained Payment for PayPal.

Please do consider supporting it, we really need it.

**Comment by sealemar** [comment]
I second that. Adaptive and Chained Payments are super useful and the community is looking forward to that support

**Comment by amirpervaiz086** [comment]
@Blahartinger thanx man. P2P sdk is also on android? Because our app is in both iOS and android natively. Btw, MoneyHoleClient looks great. I am gonna dig it out. Please let me know about android SDK.

thanx

**Comment by Blahartinger** [comment]
Its not currently, though I can port it pretty quickly to android or so for you. It's essentially a white label version of what we use to power our apps on iOS and Android. Feel free to email me to continue this dialogue if you want so we don't hijack this thread. Bryan@surpriise.me

**Comment by MichaelATang** [comment]
Hey guys, i have an ios app that requires that userA pays UserB, and i get a small commission from the transaction, i think chained payments would assist in that use case but since its not available does anyone have a solution/suggestion on how to go about doing that?

**Comment by arundevma** [comment]
Currently for IOS app you can implement this only by two transaction and separate fee(0.3$ + x%) will be charged. 

Better use Stripe. They have this feature where in one transaction you can split it between two accounts and only one fee will be charged.  Disadvantage of Stripe is that it support only few countries(US, canada, Australia,Europe) compared to PayPal. 

If Paypal modify their light box or mini browser mode in adaptive payment to support safari browser, we could implement chained or parallel payment in IOS.
See my comments above.

**Comment by amirpervaiz086** [comment]
Hi guys,
  I have to implement same p2p payment. I did research and have decided to implement adaptive chained method on server side. I will get url and open it in web view on iOS so user can approve and send that status to server to send payment to user B. there is another way using native SDK. you can pay using native SDK and send status to server and in turn server will transfer to user B.

let me know when you think?

**Comment by amirpervaiz086** [comment]
I would better go for first option so it will be handled by server side and user will checkout using paypal to send to user B.

**Comment by JJGEight** [comment]
Any updates on support?? I'm in serious need of this feature!

**Comment by amirpervaiz086** [comment]
yes please see above messages. we have explained how it can be achieved in iOS or android.

**Comment by amirpervaiz086** [comment]
we should use paypal dedicated checkout method unless paypal does not expose REST API's.

**Comment by MichaelATang** [comment]
Well i used the MPL, and since i am using swift, i had to spend a little time converting the example from objective c. I used the parallel payment feature to do two payments, once USERA does a payment it lists two recipients, USERB for comission and USERC for rest of payment

**Comment by arundevma** [comment]
@MichaelATang  But MPL does not support credit cards right?.

**Comment by Gaudon** [comment]
Wait the MPL doesn't support credit card payments?!

**Comment by MichaelATang** [comment]
@arundevma well according to this https://stackoverflow.com/questions/23712348/paypal-payments-with-both-paypal-accounts-and-credit-cards-on-ios-as-a-third-par
you should be able to accept/process credit cards with PayPal by using PayPalConfiguration.acceptCreditCards property. But since i am not in the US i don't think they would 
allow me to use this functionality

**Comment by Gaudon** [comment]
I'm also not in the US (Canada), but I feel like if this is a feature that works for the US it should work mostly everywhere.. If not I guess I'll have to go with Stripe connect.

**Comment by MichaelATang** [comment]
lol well i am in South America and Stripe is not an option nor Braintree

**Comment by garfieldmypet** [comment]
HI, PayPal Support,

Any update on IOS SDK supporting chained payment? I am sorry if I keep bugging PayPal but we really need the feature , and yet we didn't get much respond on the progress since. 

It is really important for many of us here.

**Comment by mamaral** [comment]
@garfieldmypet I wouldn't count on it happening any time soon - this issue has been open for a few years now and there's been very little in terms of response/updates.

**Comment by garfieldmypet** [comment]
Hi, @mamaral 

Is there any way we can request PayPal to bump up the priority for this feature. The reason I am saying this is because MPL is getting old and dated and we don't have any alternative solution right now.

Even the REST API doesn't support chained payment, will it even be supported or can someone from PayPal at least let us know if it is on PayPal development roadmap to continue to support chained payment in the future in their REST API.

**Comment by neilco** [comment]
@garfieldmypet This is the crux of this issue:

> Even the REST API doesn't support chained payment

The PayPal SDK can't support the Adaptive Platform APIs until the REST API ports them over from the Classic API. The pressure needs to be applied to the REST API team. I'm sure the SDK team are doing their level best to escalating this issue. 

In the interim, it would be rather nice if the SDK team considered open sourcing the iOS MPL so that the community can help with modernising the branding, UI, and iOS SDK support. At least until the REST API has support for something similar to the Adaptive Platform APIs and we can properly deprecate the MPL. 

That said, the Adaptive Platform API _is_ well documented so there's nothing stopping anyone from writing their own MPL.

**Comment by mamaral** [comment]
@garfieldmypet Don't ask me, I unfortunately have no idea.

**Comment by amirpervaiz086** [comment]
Hi guys,
 I am able to implement chained payment on iOS. I used web view for it. I will share details and app links once its published. I have managing all payments and calculations on server side so its quite easy to implement chained and let user to Pay to other user directly.

**Comment by garfieldmypet** [comment]
Hi, @abhijeetr 

Why is this issue closed? Since the feature is not available yet, how can we raise our request in the future for chained payment.

If I want to comment on the REST API issue, where should I request it, can you at least give me a link.

As a developer, I am just very very sad. I think I just have to move on.

**Comment by abhijeetr** [comment]
Extremely sorry @garfieldmypet. The issue got closed accidentally and I have re-opened it. We totally understand the limitations caused by lack of this feature and want to resolve this issue. We will get back to you with a firm plan to address this.

**Comment by garfieldmypet** [comment]
Hi, @abhijeetr 

Did you feedback to PayPal development team about this issue, and the need for this feature. Did they come back to you with a plan on this feature already?

**Comment by arundevma** [comment]
Could you modify mini browser or lightbox option to support IOS UIWebView? I think that would be the quickest method to bring a temporary solution for now?  We could launch those in UIWebView inside IOS app.

**Comment by Abhishekg219** [comment]
Hello All,
I am adding Paypal chained payment to my app.
I am getting only internal server error from Paypal.
checking and posting error -1

I am stuck here for few days. Needs help from you gentleman.

**Comment by garfieldmypet** [comment]
Hi, @abhijeetr,

You stated that you will get back to us. Did you get any update from PayPal team? It had been 1 month since you stated that you will get back to us.

**Comment by mvarie** [comment]
@abhijeetr can we have a reliable timeline that we can trust?

**Comment by amirpervaiz086** [comment]
Hi Guys,
  I am having some strange issue with live Paypal key. I am using MPL chained payment in one of my iOS app. There is a delegate method 'paymentLibraryExit' it is called when library is finished or exist. I was making api call to server and it was requirement to update status on server when payment is made. Everything worked perfect with Sandboxed key but when I switched to Live Paypal key, somehow this delegate 'paymentLibraryExit' is not calling my function where I was updating status to server. What could be possible issue? I assume if it works with sandboxed it should have been working with live as well. Any help would be appreciated.
thanx

**Comment by nextbigidea** [comment]
@amirpervaiz086 did you ever publish the solution you mentioned above?

**Comment by KeelinDevenney** [comment]
Hi guys, I was brought here by the lack of support for adaptive payments in the mobile sdk. Reading through the comments Ive never seen an issue take as long. The new sdk is fantastic by the way. It works really well, especially the card reader. But to only upgrade half the sdk and leave the rest is a nightmare. Theres so much technical debt there. Are we still blaming the "REST API" team for not updating their end. Or is there just not enough requests for this feature to push it up the priority queue. Ive noticed someone mention a moneyhole project above so I will be looking into that for now. I look forward to returning to this form to see if theres any update.

**Comment by mslinn** [comment]
This feature is very important. PayPal has been non-responsive. Unfortunately, we are forced to use Stripe until such time as PayPal provides this feature.

**Comment by mandloiamit** [comment]
I need a help regarding paypal transaction payment method. I am working on an application and I am facing problem in "Adaptive Chained Payment" using Paypal.

I have two methods to implement that are:
* MPL library(2.1.7)
* Rest API

Can you please suggest me which would be the best and effective method for applying it on both Android and iOS application.

Also I would like to know if there is any sample or demo code for "Chain payment" that would be more helpful to me to understand.

**Comment by ppmtscory** [comment]
If you needed Adaptive Payment functionality, you'll need to use MPL.  There should be a sample app included with MPL and also docs located [here](https://developer.paypal.com/docs/classic/mobile/gs_MPL/).

**Comment by braebot** [comment]
This issue is now at https://github.com/paypal/PayPal-REST-API-issues/issues/2. Follow this new issue to receive updates.
